src/MongoDB.Driver/Encryption/KmsProvidersHelper.cs

```
@@ -38,6 +38,24 @@ public static void EnsureKmsProvidersAreValid(IReadOnlyDictionary<string, IReadO
             }
         }
 
+        public static void EnsureKmsProvidersTlsSettingsAreValid(IReadOnlyDictionary<string, SslSettings> kmsProviderTlsSettings)
+        {
+            if (kmsProviderTlsSettings == null)
+            {
+                return;
+            }
+
+            foreach (var kmsProviderTlsSetting in kmsProviderTlsSettings)
+            {
+                var kmsProviderTlsSettingValue = Ensure.IsNotNull(kmsProviderTlsSetting.Value, nameof(kmsProviderTlsSetting.Value));
+
+                if (kmsProviderTlsSettingValue.ServerCertificateValidationCallback != null) // tlsInsecure
```

the spec says:

    Drivers SHOULD raise an error if insecure TLS options are set. The error MUST contain the message "Insecure TLS options prohibited". This includes options equivalent to the following URI options:

    tlsInsecure
    tlsAllowInvalidCertificates
    tlsAllowInvalidHostnames
    tlsDisableOCSPEndpointCheck
    tlsDisableCertificateRevocationCheck

In c# we supports only:

* tlsInsecure => via allowing anything in ServerCertificateValidationCallback. Since we don't have control over what user can place inside, I think it's better to not allow specifying this callback at all
* tlsDisableCertificateRevocationCheck => the default value in c# driver is tlsDisableCertificateRevocationCheck=true, so I think we should avoid this validation
* I think it probably makes sense to disable configuring SslSettings.ClientCertificateSelectionCallback even thought it's not required directly, thoughts?

---

Also, the spec has this requirement:

    Drivers MUST raise an error if the TLS options are set to disable TLS. The error MUST contain the message "TLS is required".

and `SslSettings.EnabledSslProtocols` has an option SslProtocols.None, but based on this doc https://docs.microsoft.com/en-us/dotnet/api/system.security.authentication.sslprotocols?view=net-5.0#fields it's still `Tls` option, so I didn't disable it

---

The doc says:

>SslProtocols.None  => Allows the operating system to choose the best protocol to use, and to block protocols that are not secure. 

I didn't 100% understand what this means (but I inclined that TLS is presented in any case), but you say that it means that TLS can be disabled in this case? If so, I will change it in the next commit

---

Your interpretation is correct. `SslProtocols.None` is not what I expected and neither is `SslProtocols.Default`.